### PR TITLE
[receiver/activedirectoryds] Add Active Directory Domain Services metrics receiver

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
Adds the `active_directory_ds` metrics receiver, which scrapes metrics about a running  [Active Directory Domain Services](https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/get-started/virtual-dc/active-directory-domain-services-overview) instance. 

**Link to tracking Issue:** #8833

**Testing:**
- unit tests
- integration test
  - Note about this integration test: It is currently infeasible to run it in GitHub actions. In order to run the receiver, a domain controller must be installed and running on the machine, which requires a machine restart (there doesn't seem to be a way to do this in GitHub actions). Otherwise, the performance counters have no associated instances, and the receiver fails to start. I have verified manually that this test does indeed work on a Windows server 2019 instance.

**Documentation:**
- added README

### CODE REVIEW

**Comment by djaglowski** [comment]
If I want to make a change to this receiver, what should I do to validate that I have not broken something?

**Comment by BinaryFissionGames** [comment]
@djaglowski On the system you want to test on (Windows Server 2012 R2+) -

Install go + git
Install active directory and a test forest:
```pwsh
Install-windowsfeature -name AD-Domain-Services -IncludeManagementTools

# Local administrator requires a password for installation, otherwise the next step will fail,
# so set the Admin password if necessary:
# net user Administrator P@ssw0rd!

# Install the forest; Note that the -SafeModeAdministratorPassword is set to P@ssw0rd! in this instance, and isn't secure
Install-ADDSForest -InstallDns -DomainName test.example.com -DomainNetBIOSName TEST -SafeModeAdministratorPassword (ConvertTo-SecureString -String "P@ssw0rd!" -AsPlainText -Force) -NoRebootOnCompletion -Force
```

After this step, a reboot is required. The `-NoRebootOnCompletion` flag in that last command can be removed for the system to automatically reboot, but either way, active directory will not start until the system is rebooted.

After reboot, the integration test can be run with go test at the root of the cloned repo:
```pwsh
go test -tags=integration github.com/open-telemetry/opentelemetry-collector-contrib/receiver/activedirectorydsreceiver/...
```

---

As explained above, the problem is that reboot step. I've looked into windows containers, but it has a similar issue.

The only way I can think is possibly running a VM to test, but that has a few issues:

1. I'm not sure if it's technically feasible to run a Windows VM within Github Actions
2. It will likely add a decent chunk of time to the running of actions.
3. I don't know the licensing implications of running a Windows VM in CI like that.

**Review by djaglowski — COMMENTED** [review]
@BinaryFissionGames, thanks for the detailed description of how to create a test system.

It seems that a true integration test is not realistic for this receiver. However, I would like to explore the idea of creating a simple simulator, rather than relying on "manual" mocking and testing. Please see other comments.

File: receiver/activedirectorydsreceiver/counters.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	DSSearchSubOperations                  winperfcounters.PerfCounterWatcher
+	DSSecurityDescripterSubOperations      winperfcounters.PerfCounterWatcher
+	DSThreadsInUse                         winperfcounters.PerfCounterWatcher
+	LDAPClientSessions                     winperfcounters.PerfCounterWatcher
+	LDAPBindTime                           winperfcounters.PerfCounterWatcher
+	LDAPSuccessfulBinds                    winperfcounters.PerfCounterWatcher
+	LDAPSearches                           winperfcounters.PerfCounterWatcher
+}
+
+func getWatchers() (*watchers, error) {
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply receiver/activedirectorydsreceiver/counters.go:57]: This is a great idea, let me research about crawling counters and see what I can come up with.

The counters actually do end up being created before the required restart (I think right away when the feature is enabled), so we could probably even scrape counter definitions in CI to add a good amount of confidence.

File: receiver/activedirectorydsreceiver/scraper.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	return pdata.Metrics(a.mb.Emit()), nil
+}
+
+func (a *activeDirectoryDSScraper) shutdown(ctx context.Context) error {
+	if a.w != nil {
+		var err error
+
+		err = multierr.Append(err, a.w.DRAInboundBytesCompressed.Close())
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply receiver/activedirectorydsreceiver/scraper.go]: I refactored this a bit; I did some sort of middle ground here, using a map instead. I think it makes it a bit more safe in regards to closing specifically.

File: receiver/activedirectorydsreceiver/counters.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply receiver/activedirectorydsreceiver/counters.go]: Implemented this suggestion; The scraping in CI seems to work too, which is awesome.

**Review by djaglowski — APPROVED** [review]
LGTM. Thanks @BinaryFissionGames

**Pull request merged**